rust/datafusion/src/logicalplan.rs

```diff
@@ -725,36 +725,47 @@ impl fmt::Debug for LogicalPlan {
 /// Verify a given type cast can be performed
 pub fn can_coerce_from(type_into: &DataType, type_from: &DataType) -> bool {
     use self::DataType::*;
+
+    if type_from == type_into {
+        return true;
+    }
+
     match type_into {
         Int8 => match type_from {
             Int8 => true,
             _ => false,
         },
         Int16 => match type_from {
-            Int8 | Int16 | UInt8 => true,
+            Int8 | Int16 => true,
+            UInt8 => true,
             _ => false,
         },
         Int32 => match type_from {
-            Int8 | Int16 | Int32 | UInt8 | UInt16 => true,
+            Int8 | Int16 | Int32 => true,
+            UInt8 | UInt16 => true,
             _ => false,
         },
         Int64 => match type_from {
-            Int8 | Int16 | Int32 | Int64 | UInt8 | UInt16 | UInt32 => true,
+            Int8 | Int16 | Int32 | Int64 => true,
+            UInt8 | UInt16 | UInt32 => true,
             _ => false,
         },
         UInt8 => match type_from {
             UInt8 => true,
             _ => false,
         },
         UInt16 => match type_from {
+            Int8 => true,
```

Comment:
for my own curiosity, what's the expected behavior when a negative int8 is coerced into uint16?

Reply:
looking at the code, looks like it will return null for negative values. if that's the case, we should be able to support conversion from signed to unsigned with the same width as well right?

Reply:
Thanks, this is actually a bug. The goal of this can_coalesce method is to determine if DataFusion can safely cast from one type to another without any data loss, so it can add implicit casts in queries. This is convenient when the user is comparing two expressions of different types.

Separately from this, users can add any explicit CAST they want to in their query plan.